File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        n = len(nums)
        subset = []
        
        def backtrack(i):
            if i >= n:
                result.append(subset.copy())
                return
            
            subset.append(nums[i])
            backtrack(i + 1)
            subset.pop()
            backtrack(i + 1)
        
        backtrack(0)
        return result
    
    
            
            
            
            

File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList();
        List<Integer> subset = new ArrayList();
        
        backtrack(nums, subset, result, 0);
        
        return result;
        
    }
    
    public void backtrack(int[] nums, List<Integer> subset, List<List<Integer>> result, int i) {
        int n = nums.length;
        if (i >= n) {
            result.add(new ArrayList(subset));
            return;
        }
        
        subset.add(nums[i]);
        backtrack(nums, subset, result, i + 1);
        subset.remove(subset.size() - 1);
        backtrack(nums, subset, result, i + 1);
        
    }
}